### PR TITLE
Added support for Continuous environments and refactored some code

### PR DESCRIPTION
**Pull request by menkaur**
I have looked through the code responsible for reinforcement learning, and found following issues:
1. in /worker, all workers were copy pastes, with minor changes. I've refactored it as far as I could by creating worker base class and doing there resource management. I didn't do any refactoring on the Step methods
2. Greedy_policy will not work with continuous models, I've added static assert to make sure that the code will not compile if actiontype is anything but enum
3. Created continuous worker class which supports continuous environments
4. Implementation of epsilon greedy policy decays lineary, which has caused some issues in my tests (where the model failed to learn if I modified decay rate). I've added ExponentiallyDecayingGreedyPolicy, which is easier to use

Commited code can be tested with

```
int main()
{
    const bool testDiscreteLearning=false;

    // Set up the network.




    const double decayRate = 1e-5;

    if(testDiscreteLearning)
    {
        const int stepLimit = 300;

        TrainingConfig config;
        config.StepSize() = 0.01;
        config.Discount() = .9;
        config.TargetNetworkSyncInterval() = 100;
        config.ExplorationSteps() = 100;
        config.DoubleQLearning() = true;
        config.StepLimit() = stepLimit;

        FFN <MeanSquaredError<>, GaussianInitialization> model(MeanSquaredError<>(), GaussianInitialization(0, 0.01));
        model.Add<Linear<>>(CartPole::State::dimension, 128);
        //model.Add<TanHLayer<>>();
        model.Add<LeakyReLU<>>();
        model.Add<Linear<>>(128, 128);
        //model.Add<TanHLayer<>>();
        model.Add<LeakyReLU<>>();
        model.Add<Linear<>>(128, 128);
        //model.Add<TanHLayer<>>();
        model.Add<LeakyReLU<>>();
        //model.Add<ReLULayer<>>();
        model.Add<Linear<>>(128, CartPole::Action::size);

        /*AggregatedPolicy <ExponentiallyDecayingGreedyPolicy<CartPole>> policy({ExponentiallyDecayingGreedyPolicy<CartPole>(0.7, 0,decayRate),
                                                          ExponentiallyDecayingGreedyPolicy<CartPole>(0.7, 0.01,decayRate),
                                                          ExponentiallyDecayingGreedyPolicy<CartPole>(0.7, 0.5,decayRate)},
                                                         arma::colvec("0.4 0.3 0.3"));*/
        ExponentiallyDecayingGreedyPolicy<CartPole> policy(0.7, .01,decayRate);

        OneStepQLearning<CartPole, decltype(model), ens::VanillaUpdate, decltype(policy)>
                agent(std::move(config), std::move(model), std::move(policy));

        arma::vec returns(100, arma::fill::zeros);
        size_t position = 0;
        size_t episode = 0;
        int executionPosition = -1;

        auto measure = [&](double episodeReturn) {
            returns[position++] = episodeReturn;
            position = position % returns.n_elem;
            episode++;

            auto mean = arma::mean(returns);

            if (mean > stepLimit * .95 && episode > returns.n_elem) {
                std::cout << "returning because mean=" << mean << std::endl;
                return true;
            }

            std::cout << "[" << executionPosition << "," << episode << "]"
                      << "; Episode Return: " << episodeReturn
                      << "; Average Return: " << mean
                      << std::endl;

            return false;
        };

        agent.Train(measure);
    }
    else
    {
        const int stepLimit = 100;

        TrainingConfig config;
        config.StepSize() = 0.001;
        config.Discount() = .9;
        config.TargetNetworkSyncInterval() = 100;
        config.ExplorationSteps() = 100;
        config.DoubleQLearning() = true;
        config.StepLimit() = stepLimit;
        config.NumWorkers()=1;

        typedef ContinuousMountainCar workType;
        FFN <MeanSquaredError<>, GaussianInitialization> model(MeanSquaredError<>(), GaussianInitialization(0, .1));
        model.Add<Linear<>>(workType::State::dimension, 128);
        //model.Add<TanHLayer<>>();
        model.Add<LeakyReLU<>>();
        model.Add<Linear<>>(128, 128);
        //model.Add<TanHLayer<>>();
        model.Add<LeakyReLU<>>();
        model.Add<Linear<>>(128, 128);
        //model.Add<TanHLayer<>>();
        model.Add<LeakyReLU<>>();
        //model.Add<ReLULayer<>>();
        model.Add<Linear<>>(128, workType::Action::size);



        ContinuousGreedyPolicy<workType> policy(1, .01,decayRate);

        OneStepContinuousQLearning<workType, decltype(model), ens::VanillaUpdate, decltype(policy)>
                agent(std::move(config), std::move(model), std::move(policy));

        arma::vec returns(100, arma::fill::zeros);
        size_t position = 0;
        size_t episode = 0;
        int executionPosition = -1;

        auto measure = [&](double episodeReturn) {
            returns[position++] = episodeReturn;
            position = position % returns.n_elem;
            episode++;

            auto mean = arma::mean(returns);

            if (mean > stepLimit * .95 && episode > returns.n_elem) {
                std::cout << "returning because mean=" << mean << std::endl;
                return true;
            }

            std::cout << "[" << executionPosition << "," << episode << "]"
                      << "; Episode Return: " << episodeReturn
                      << "; Average Return: " << mean
                      << std::endl;

            return false;
        };

        agent.Train(measure);
    }
}
```

### CODE REVIEW

File: src/mlpack/methods/reinforcement_learning/policy/exp_decay_greedy_policy.hpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/**
+ * @file exp_decaay_greedy_policy.hpp
```
Comment (birm) [inline src/mlpack/methods/reinforcement_learning/policy/exp_decay_greedy_policy.hpp:2]:
```suggestion
 * @file methods/reinforcement_learning/policy/exp_decay_greedy_policy.hpp
```
this should fix the doc build

File: src/mlpack/methods/reinforcement_learning/policy/aggregated_policy.hpp [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /**
- * @file methods/reinforcement_learning/policy/aggregated_policy.hpp
+ * @file aggregated_policy.hpp
```
Comment (birm) [inline src/mlpack/methods/reinforcement_learning/policy/aggregated_policy.hpp:2]: Do you have a reason for this change?

File: src/mlpack/methods/reinforcement_learning/environment/continuous_mountain_car.hpp [hunk 3]
```diff
@@ -140,7 +169,9 @@ class ContinuousMountainCar
     double force = math::ClampRange(action.action[0], -1.0, 1.0);
 
     // Update states.
-    nextState.Velocity() = state.Velocity() + force * duration - 0.0025 *
+    nextState.Velocity() = state.Velocity() + force * duration -
+            //0.0025 *
+            0.0014*//for the simulation to be theoretically able to reach the end goal, this value must be less than the value of duration
```
Comment (birm) [inline src/mlpack/methods/reinforcement_learning/environment/continuous_mountain_car.hpp:174]: If we're hard coding this value, does it make sense to have a comment like this?

File: src/mlpack/methods/reinforcement_learning/policy/aggregated_policy.hpp (hunk 2 above)
Comment (menkaur) [reply src/mlpack/methods/reinforcement_learning/policy/aggregated_policy.hpp:2]: no, it is identical to the original

**Comment by menkaur** [comment]
How can I check what's wrong with the LaTeX Documentation Checks?

**Comment by menkaur** [comment]
Ok, so the compilation keeps failing because of an error in 

`/home/vsts/work/1/s/src/mlpack/tests/rl_components_test.cpp:92:`

this is not something that I modified, so I'll just wait for any input you guys have

File: src/mlpack/methods/reinforcement_learning/environment/continuous_mountain_car.hpp [hunk 4]
```diff
@@ -86,7 +87,35 @@ class ContinuousMountainCar
   {
     double action[1];
     // Storing degree of freedom
-    const int size = 1;
+    static constexpr int size = 1;
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/environment/continuous_mountain_car.hpp:90]: Hm, I think that should work, but looking at the build error:

`rl_components_test.cpp:92: undefined reference to mlpack::rl::ContinuousMountainCar::Action::size'`

Can you just use `const int size = 1;` to see if that would fix the issue?

**Comment by zoq** [comment]
Let's keep this one open.

**Pull request closed without merging**